### PR TITLE
enc, change salt handling, way 1

### PR DESCRIPTION
**Pull request by EasySec**
This is an attempt to workaround the 'curse' of the salt handling using the `enc` command :
The salt is not always stored with the encrypted data.
As could be expected, it is stored when it was generated on the fly (not supplied with args), but not only.
It is stored when no key is given.
And when stored, even good IV and key fails to decrypt.
You must enter the password (but are NOT prompted for).

This first PR act as folows : 
If encrypting generate a salt IF NOT supplied using args.
If decrypting use salt from input BIO IF NOT supplied using args.

see also #4487, #2083

### CODE REVIEW

**Comment by richsalz** [comment]
This is not strictly a 1.1.1 bug or feature-add, so I am removing the 1.1.1 milestone.

**Review by t8m — COMMENTED** [review]
I think this should be merged. Could you please adjust the enc manual page to clarify how salt is handled with your patch?

**Comment by t8m** [comment]
We are sorry for neglecting this for so long.

**Comment by EasySec** [comment]
> 
> 
> Could you please adjust the enc manual page to clarify how salt is handled with your patch?

I made a quick return on that matter today. What about the content of my new commit ? Do you think it's clear enough ?

File: apps/enc.c [hunk 1]
```diff
@@ -429,51 +429,62 @@ int enc_main(int argc, char **argv)
     }
 
     if (cipher != NULL) {
-        /*
-         * Note that str is NULL if a key was passed on the command line, so
-         * we get no salt in that case. Is this a bug?
-         */
-        if (str != NULL) {
+
```
Comment (t8m) [inline apps/enc.c:432]: Please drop this empty line.

File: apps/enc.c [hunk 2]
```diff
@@ -429,51 +429,62 @@ int enc_main(int argc, char **argv)
...
-                    if ((printkey != 2)
-                        && (BIO_write(wbio, magic,
-                                      sizeof(magic) - 1) != sizeof(magic) - 1
-                            || BIO_write(wbio,
-                                         (char *)salt,
-                                         sizeof(salt)) != sizeof(salt))) {
-                        BIO_printf(bio_err, "error writing output file\n");
-                        goto end;
+                } else {    /* decryption */
+
```
Comment (t8m) [inline apps/enc.c:470]: Please drop this empty line

File: apps/enc.c [hunk 3]
```diff
@@ -429,51 +429,62 @@ int enc_main(int argc, char **argv)
...
+                            goto end;
+                        }
+                        if (memcmp(mbuf, magic, sizeof(mbuf)) == 0) { /* file IS salted */
+                            if (BIO_read(rbio, salt,
+                                         sizeof(salt)) != sizeof(salt)) {
+                                BIO_printf(bio_err, "error reading input file\n");
+                                goto end;
+                            }
+                        }
+                        else { /* file is NOT salted, NO salt available */
```
Comment (t8m) [inline apps/enc.c:483]: Else should be placed immediately after the `}`

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few code formatting nits to fix.

**Comment by t8m** [comment]
> I made a quick return on that matter today. What about the content of my new commit ? Do you think it's clear enough ?

I think it is OK.

**Comment by EasySec** [comment]
> 
> 
> Just a few code formatting nits to fix.

Yes, sorry. Done.

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by mattcaswell** [comment]
Pushed!! Thanks for your contribution. It only took 3.5 years to merge this!

Prior to pushing I made a trivial edit to the documentation part of this PR to remove some spurious end of line whitespace.

**Pull request closed without merging**